### DISCUSSION TITLE
Scalable way to manage navigator and screen names?

### DISCUSSION THREAD

**Discussion by emeraldsanto** [react-navigation/react-navigation#10712] [Q&A] [answered]
### Summary

Hello 👋🏻 

I've been using `react-navigation` for around 3.5 years and I always end up running into this problem where the amount of screens grows but the strategy for managing them does not adapt.

I always default to the strategy of having an enum for navigators and one for screens which, at scale, feels incredibly naive.

```typescript
enum Screen {
  OnboardingCredentials = 'onboarding-credentials-screen',
  // ...
  PreferencesAlerts = 'preferences-alerts-screen',
  // ...
  SomeOther = 'some-other-screen',
}

interface OnboardingStackParameters {
  [Screen.OnboardingCredentials]: { referrer: string },
  // ...

  [Screen.SomeOtherScreen]: { whatever: boolean }
}

interface PreferencesStackParameters {
  [Screen.PreferencesAlert]: undefined,
  // ...
}

// ...

navigation.navigate(Navigator.Onboarding, { screen: Screen.OnboardingCredentials }); 
```

I get that navigating purely by path does not provide as much flexibility as the above given we can't specify whether to `replace` or `reset`, for example, but I doubt this is the best we can collectively come up with?

Any other recommended approaches?

**Comment by satya164** [comment, marked as answer]
Why have the enums? With proper annotation, the navigation methods are already typesafe, so the enums don't provide any additional value.

**Reply by emeraldsanto** [reply]
I always figured that to be better for to how often each screen has to be referenced in different places (no magic strings) 😵

**Reply by satya164** [reply]
They are typesafe and provide the same DX as an enum (autocompletion, error-checking etc.) without having to add an extra import and having to define the enum separately. They aren't magic strings since what you write is already in the type system.

**Reply by emeraldsanto** [reply]
That may be the case for most `react-navigation` APIs but I noticed `Stack.Screen` props allow any strings, even when created like so `createNativeStackNavigator<OnboardingStackParameters & ParamListBase>`.

**Reply by satya164** [reply]
The `& ParamListBase` shouldn't be there.

**Reply by emeraldsanto** [reply]
If I could omit it, I would 😬 

```
Type 'OnboardingStackParameters' does not satisfy the constraint 'ParamListBase'.
  Index signature for type 'string' is missing in type 'OnboardingStackParameters'.
```

It's either that or `interface OnboardingStackParameters extends ParamListBase {}`

**Reply by satya164** [reply]
@emeraldsanto they can't be an interface. You need to use `type` like in the docs.

**Reply by emeraldsanto** [reply]
TIL, I just thought that was a style-guide choice since there's no mention of why `type` is needed in the documentation. I've just tried it out and it did work as expected 🤯
